osvr_tracked_device.h

```diff
@@ -340,107 +331,49 @@ bool OSVRTrackedDevice::IsDisplayRealDisplay()
-    // This is similar to glFrustum but not quite the same!
-    // idx = 1 / (right - left)
-    // idy = 1 / (bottom - top)
-    // idz = 1 / (far - near)
-    // sx = right + left
-    // sy = bottom + top
-    // eye(0, 0) = 2 * idx = 2 / (right - left)
-    // eye(1, 1) = 2 * idy = 2 / (bottom - top)
-    // eye(0, 2) = sx * idx = (right + left) / (right - left)
-    // eye(1, 2) = sy * idy = (bottom + top) / (bottom - top)
-    // eye(2, 2) = -far * idz = -far / (far - near)
-    // eye(2, 3) = -(far * near * idz) = -(far * near) / (far - near)
-    // eye(3, 2) = -1
-
-    const double dx = 2.0 / eye_matrix(0, 0); // (right - left)
-    const double sx = eye_matrix(0, 2) * dx; // (right + left)
-    *right = static_cast<float>((sx + dx) / 2.0); // right
-    *left = static_cast<float>(sx - *right);
-
-    const double dy = 2.0 / eye_matrix(1, 1); // (bottom - top)
-    const double sy = eye_matrix(1, 2) * dy; // (bottom + top)
-    *bottom = static_cast<float>((sy + dy) / 2.0); // bottom
-    *top = static_cast<float>(sy - *bottom);
+    // SteamVR expects top and bottom to be swapped!
+    osvr::clientkit::ProjectionClippingPlanes pl = m_DisplayConfig.getViewer(0).getEye(eye).getSurface(0).getProjectionClippingPlanes();
+    *left = static_cast<float>(pl.left);
+    *right = static_cast<float>(pl.right);
+    *bottom = static_cast<float>(pl.top); // SWAPPED
+    *top = static_cast<float>(pl.bottom); // SWAPPED
 }
 
 vr::HmdMatrix34_t OSVRTrackedDevice::GetHeadFromEyePose(vr::Hmd_Eye eye)
 {
+    OSVR_Pose3 headPose, eyePose;
+    vr::HmdMatrix34_t matrix;
 
-    // Rotate per the display configuration
-    const double horiz_fov = m_DisplayConfiguration->getHorizontalFOVRadians();
-    const double overlap = m_DisplayConfiguration->getOverlapPercent() * horiz_fov;
-    double angle = (horiz_fov - overlap) / 2.0;
-    if (vr::Eye_Right == eye) {
-        angle *= -1.0;
+    if(m_DisplayConfig.getViewer(0).getPose(headPose) != true) {
+        logger_->Log("OSVRTrackedDevice::GetHeadFromEyePose(): Unable to get head pose!");
     }
-    const Eigen::Affine3d rotation = Eigen::Affine3d(Eigen::AngleAxisd(angle, Eigen::Vector3d(0, 1, 0)));
-
-    // Translate along x-axis by half the interpupillary distance
-    double eye_translation = m_DisplayConfiguration->getIPDMeters() / 2.0;
-    if (vr::Eye_Left == eye) {
-        eye_translation *= -1.0;
+    if(m_DisplayConfig.getViewer(0).getEye(eye).getPose(eyePose) != true) {
+        logger_->Log("OSVRTrackedDevice::GetHeadFromEyePose(): Unable to get eye pose!");
     }
-    const Eigen::Affine3d translation(Eigen::Translation3d(Eigen::Vector3d(eye_translation, 0.0, 0.0)));
-    // Eye matrix
-    const Matrix34f mat = (translation * rotation).matrix().block(0,0,3,4).cast<float>();
 
-    vr::HmdMatrix34_t matrix;
-    map(matrix) = (mat);
+    /// @todo FIXME Implement this (needs to return eye pose in head space) -- returns identity for now
+    map(matrix) = Matrix34f::Identity();
```

Comment:
I'm aware — the documentation at https://github.com/ValveSoftware/openvr/blob/9a27011472d69271cd3924efa1a31d72435ac248/headers/openvr_driver.h#L897 indicates that this should be the transformation from eye space to head space, so hTe (which is what my implementation produces) would be correct (and my earlier source code comment was not).